ffmpeg/ffmpeg.go

```diff
@@ -492,6 +507,89 @@ func Transcode3(input *TranscodeOptionsIn, ps []TranscodeOptions) (*TranscodeRes
 	return t.Transcode(input, ps)
 }
 
+type CodingSizeLimit struct {
+	WidthMin, HeightMin int
+	WidthMax, HeightMax int
+}
+
+type Size struct {
+	W, H int
+}
+
+func (s *Size) Valid(l *CodingSizeLimit) bool {
+	if s.W < l.WidthMin || s.W > l.WidthMax || s.H < l.HeightMin || s.H > l.HeightMax {
+		return false
+	}
+	return true
+}
+
+func clamp(val, min, max int) int {
+	if val <= min {
+		return min
+	}
+	if val >= max {
+		return max
+	}
+	return val
+}
+
+func (l *CodingSizeLimit) Clamp(p *VideoProfile, format MediaFormatInfo) error {
+	w, h, err := VideoProfileResolution(*p)
+	if err != nil {
+		return err
+	}
+	// detect correct rotation
+	outputAr := float32(w) / float32(h)
+	inputAr := float32(format.Width) / float32(format.Height)
+	if (inputAr > 1.0) != (outputAr > 1.0) {
+		// comparing landscape to portrait, apply rotate on chosen resolution
+		w, h = h, w
+	}
+	// Adjust to minimal encode dimensions keeping aspect ratio
+
+	var adjustedWidth, adjustedHeight Size
+	adjustedWidth.W = clamp(w, l.WidthMin, l.WidthMax)
+	adjustedWidth.H = format.ScaledHeight(adjustedWidth.W)
+	adjustedHeight.H = clamp(h, l.HeightMin, l.HeightMax)
+	adjustedHeight.W = format.ScaledWidth(adjustedHeight.H)
+	if adjustedWidth.Valid(l) {
+		p.Resolution = fmt.Sprintf("%dx%d", adjustedWidth.W, adjustedWidth.H)
+		return nil
+	}
+	if adjustedHeight.Valid(l) {
+		p.Resolution = fmt.Sprintf("%dx%d", adjustedHeight.W, adjustedHeight.H)
+		return nil
+	}
+	return fmt.Errorf("profile %dx%d size out of bounds %dx%d-%dx%d", w, h, l.WidthMin, l.WidthMin, l.WidthMax, l.HeightMax)
+}
+
+// 7th Gen NVENC limits:
+var nvidiaCodecSizeLimts = map[VideoCodec]CodingSizeLimit{
+	H264: {146, 50, 4096, 4096},
+	H265: {132, 40, 8192, 8192},
+}
+
+func ensureEncoderLimits(outputs []TranscodeOptions, format MediaFormatInfo) error {
+	// not using range to be able to make inplace modifications to outputs elements
+	for i := 0; i < len(outputs); i++ {
+		if outputs[i].Accel == Nvidia {
+			limits, haveLimits := nvidiaCodecSizeLimts[outputs[i].Profile.Encoder]
+			resolutionSpecified := outputs[i].Profile.Resolution != ""
+			// Sometimes rendition Resolution is not specified. We skip this rendition.
```

Comment:
@thomshutt Some unit tests are failing. Do you want me to address them before merging?